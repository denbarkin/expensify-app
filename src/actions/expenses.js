import uuid from 'uuid';
import database from '../firebase/firebase'; // imports default exports.

// Add Expense
export const addExpense = (
    {
        description = '', 
        note = '', 
        amount = 0, 
        createdAt = 0
    } = {}
    ) => {
    return({
        type: 'ADD_EXPENSE',
        expense :{
            id : uuid(),
            description : description,
            note,
            amount,
            createdAt
        }
    })
}

export const addExpenseV2 = (expense) => {
    return({
        type: 'ADD_EXPENSE',
        expense 
    })
}

// Asynchronous Redux Action with Function.
// "redux-thunk": "^2.3.0", is used to send function to Redux Action.
export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData; //destructure

        const expense = {description, note, amount,createdAt};
        const uid = getState().auth.uid;

        return database.ref(`users/${uid}/expenses`).push(expense)
            .then((ref) => {
                dispatch(addExpenseV2({
                    id : ref.key,
                    ...expense
                }))
            })
            .catch((e) => {
                console.log("Error:", e.message); 
            })
    }
}
// Remove Expense
export const removeExpense = ({id} = {}) => {
    return({
        type : 'REMOVE_EXPENSE',
        expense : {
            id : id
        }
    })
}
// Asynch action return arrow function.
export const startRemoveExpense = ({id} = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({id}))
        })
    }
}
// Edit Expense
export const editExpense = (id, updates ) => {
    return (
        {
            type: 'EDIT_EXPENSE',
            id,
            updates
        }
    );
}
// Asynch action return arrow function.
export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id,updates));
        })
    }
}

export const setExpenses = (expenses) => (
    {
        type: 'SET_EXPENSES',
        expenses
    } 
)

// Asynchronous Redux Action with Function.
// 1. Fetch All expense data once.
// 2. Parse the data into array
// 3. Dispatch setExpenses to update store.
export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        // 1. Fetch All expense data once.
        return database.ref(`users/${uid}/expenses`).once('value')
        .then((snapshot) => {
            // 2. Parse the data into array
            const expenses = [];
            snapshot.forEach(childsnapshot => {
                expenses.push({
                    id : childsnapshot.key,
                    ...childsnapshot.val()
                })
            });
            // 3. Dispatch setExpenses to update store.
            dispatch(setExpenses(expenses));

        }).catch((error) => {
            console.log('Error Set Expenses Asynch: ', error.message);
        })
    }
}