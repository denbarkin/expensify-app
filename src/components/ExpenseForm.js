import React from "react";
import moment from "moment";
import { SingleDatePicker } from 'react-dates';


export default class ExpenseForm extends React.Component
{
    constructor(props){
        super(props)

        this.state = {
            description : props.expense ?  props.expense.description : '',
            amount : props.expense ?  props.expense.amount : 0,
            note : props.expense ?  props.expense.note : '',
            createdAt: props.expense ?  moment(props.expense.createdAt) : moment(),
            calenderFocused : false,
            error : ''
        }
    } 

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        // regex101.com for detailed regex 
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
        {   
            this.setState(() => ({amount}))
        }
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}))
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt })) 
        }  
    }

    onFocusChange = ({focused}) => {
        this.setState(() => ({ calenderFocused : focused }))
    }

    onSubmitHandler = (e) => {
        e.preventDefault();

        let error = '';
        if (!this.state.description || !this.state.amount) {
            // Error log
            error = 'Description and Amount must be valid.'
            this.setState(() => ({ error }));
        } else {
            // Clear Error Log 
            error = '';
            this.setState(() => ({error}))
            this.props.onSubmit({
                description : this.state.description,
                note : this.state.note,
                amount : this.state.amount,
                createdAt : this.state.createdAt.valueOf()
            });
        }
    }
    
    render(){
        return(
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmitHandler}>
                    <input 
                        type='text'
                        placeholder="description"
                        value={this.state.description}
                        autoFocus
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type='text'
                        placeholder='amount'
                        value = {this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker 
                        date = {this.state.createdAt}
                        onDateChange = {this.onDateChange}
                        onFocusChange = {this.onFocusChange}
                        focused = {this.state.calenderFocused}
                        numberOfMonths = {1}
                        isOutsideRange = { () => false}
                    />
                    <textarea 
                        type='text'
                        value = {this.state.note}
                        placeholder='Notes for the expense'
                        onChange={this.onNoteChange}
                    />
                    <button type='submit'>Add Expense</button>
                </form>  
            </div>
        )
    }
}
