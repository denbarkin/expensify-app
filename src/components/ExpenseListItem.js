import React from "react";
import { connect } from "react-redux";
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import moment from 'moment';

const ExpenseListItem = ({id, description, amount, createdAt}) => {
    return(
        <div>
            <Link to={`Edit/${id}`} >
                <h3>{description}</h3>
            </Link>
            <p>
                {numeral(amount).format('$0,0.00')} 
                - 
                {moment(createdAt).format('MMMM Do, YYYY')}
            </p>
        </div>
    );
};

export default connect()(ExpenseListItem); // HOC component