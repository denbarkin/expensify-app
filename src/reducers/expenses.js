const expenseReducerDefaultState = [];

export default  (state = expenseReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            //return state.concat(action.expense); // returns new object.
            return [
                ...state,   // spread operator returns new array
                action.expense
            ]
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) =>  id !== action.expense.id); 

        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates  // Updates the amount in JSON with spread operator.
                    }
                }
                else{
                    return expense;
                }
            })
        default:
            return state;
    }
}