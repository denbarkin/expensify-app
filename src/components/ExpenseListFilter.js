import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByDate, sortByAmount, setStartDateFilter,setEndDateFilter } from "../actions/filters";
import { DateRangePicker } from "react-dates";

class ExpenseListFilter extends React.Component
{
    state = {
        focusedInput: null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.dispatch(setStartDateFilter(startDate));
        this.props.dispatch(setEndDateFilter(endDate));
    }

    render(){
        return(
            <div>
                <input value={this.props.filters.text} onChange={ (e) => {
                        // e.target.style.backgroundColor = '#200355';
                        this.props.dispatch(setTextFilter({text:e.target.value}))
                    }
                }>
                </input>
                <select value={this.props.filters.sortBy} onChange={ (e) => {
                    e.defaultPrevented = true;
                    if (e.target.value === 'date') {
                        this.props.dispatch(sortByDate())
                    } else {
                        this.props.dispatch(sortByAmount())
                    }
                }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                  //  startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                    endDate={this.props.filters.endDate} 
                    //endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={focusedInput => this.setState({ focusedInput })} // PropTypes.func.isRequired,
                    showClearDates = {true}
                    isOutsideRange = {() => false}
                 />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters : state.filters
    }
}

export default connect(mapStateToProps)(ExpenseListFilter);