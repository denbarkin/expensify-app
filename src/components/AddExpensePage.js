import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startAddExpense } from '../actions/expenses';
import React, { Component } from 'react'

export class AddExpensePage extends Component {
  
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    }

    render() {
    return (
      <div>
        <h1>Add Expense Form</h1>
        <ExpenseForm onSubmit = { this.onSubmit }/>  
      </div>
    )
  }
}

const mapDispatchToProp = (dispatch) => ({
    startAddExpense : (expense) => dispatch(startAddExpense(expense))
})

export default connect(undefined, mapDispatchToProp)(AddExpensePage);