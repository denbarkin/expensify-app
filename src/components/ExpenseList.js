import { connect } from "react-redux";
import React from "react";
import ExpenseListItem from "./ExpenseListItem";
import selectedExpenses from "../selectors/expenses";
import getTotalExpenses from "../selectors/expenses-total";
import getCountExpenses from "../selectors/expenses-count";
import ExpenseListFilter from "./ExpenseListFilter";
import ExpenseSummary from "./ExpenseSummary";

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        <ExpenseListFilter />
        <ExpenseSummary expenses={props.expenses} />
        {props.expenses.map((expense) => {
           return <ExpenseListItem key={expense.id} {...expense} />
        })}
    </div>
);

const mapStateToProps = (state) => {
    return {
      expenses : selectedExpenses(state.expenses,state.filters)
    };
};  

// HOC returns with connect overrided ExpenseList.
export default connect(
    mapStateToProps
  )(ExpenseList);
