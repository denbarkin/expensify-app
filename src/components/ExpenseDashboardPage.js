import React from 'react';
import ExpenseList from "./ExpenseList";

const ExpenseDashboardPage = () => {
    return(
        <div>
            <div><ExpenseList /></div>
        </div>
    )  
}

export default ExpenseDashboardPage;