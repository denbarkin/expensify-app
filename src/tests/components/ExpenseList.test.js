import React from 'react';
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

// Snapshot testing with Dynamic Components 
// passing data with props
test('should render ExpenseList', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses}/>);
    expect(wrapper).toMatchSnapshot();
})