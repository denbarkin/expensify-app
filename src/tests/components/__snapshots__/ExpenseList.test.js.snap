// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <Connect(ExpenseListFilter) />
  <ExpenseSummary
    expenses={
      Array [
        Object {
          "amount": 10,
          "createdAt": 0,
          "description": "Gum",
          "id": 1,
          "note": "just a gum",
        },
        Object {
          "amount": 1020,
          "createdAt": -345600000,
          "description": "Water Bill",
          "id": 2,
          "note": "Big bill",
        },
        Object {
          "amount": 100000,
          "createdAt": 345600000,
          "description": "New Car",
          "id": 3,
          "note": "Tesla car",
        },
      ]
    }
  />
  <Connect(ExpenseListItem)
    amount={10}
    createdAt={0}
    description="Gum"
    id={1}
    key="1"
    note="just a gum"
  />
  <Connect(ExpenseListItem)
    amount={1020}
    createdAt={-345600000}
    description="Water Bill"
    id={2}
    key="2"
    note="Big bill"
  />
  <Connect(ExpenseListItem)
    amount={100000}
    createdAt={345600000}
    description="New Car"
    id={3}
    key="3"
    note="Tesla car"
  />
</div>
`;
